### PR TITLE
Generator functions

### PR DESCRIPTION
**Pull request by andreiradu**
~~One particular issue is that calling `iterator.next(42)` transpiles to `iterator:next(42)`, which in turn means that the corresponding `coroutine.yield()` will return `iterator, 42`. I've made it so `let a = yield()` transpiles to `local _, a = coroutine.yield()` to account for that.~~
```ts
function* seq() {
    let a = yield 1;
    return a;
}
const gen = seq();
```
will transpile to:
```lua
seq = function()
    local __co = coroutine.create(function()
        local a = coroutine.yield(1);
        return a;
    end);
    __it = {next = function(___, ...)
        local __err, __value = coroutine.resume(__co, ...);
        if not __err then
            error(__value);
        end
        return {done = coroutine.status(__co) == "dead", value = __value};
    end};
    __it[Symbol.iterator] = function()
        return __it;
    end;
    return __it;
end;
local gen = seq();
```

### CODE REVIEW

File: src/LuaAST.ts [hunk 1]
```diff
@@ -813,6 +813,12 @@ export function createIdentifier(
     return expression;
 }
 
+export function createAnnonymousIdentifier(tsOriginal?: ts.Node, parent?: Node): Identifier {
+    const expression = createNode(SyntaxKind.Identifier, tsOriginal, parent) as Identifier;
+    expression.text = "___";
```
Comment (tomblind) [inline src/LuaAST.ts]: this should be 4 underscores for consistency

File: src/LuaTransformer.ts [hunk 2]
```diff
@@ -1031,6 +1030,143 @@ export class LuaTransformer {
...
+        nextBody.push(tstl.createReturnStatement([iteratorResult]));
+
+        //function(__, ...)
+        const nextFunctionDeclaration = tstl.createFunctionExpression(
+            tstl.createBlock(nextBody),
+            [tstl.createAnnonymousIdentifier()],
+            tstl.createDotsLiteral());
+
+        //__it = {next = function(__, ...)}
+        const iterator = tstl.createAssignmentStatement(
```
Comment (tomblind) [inline src/LuaTransformer.ts]: Shouldn't this be a VariableDeclarationStatement so `__it` is marked local?

File: src/LuaTransformer.ts [hunk 3]
```diff
@@ -1031,6 +1030,143 @@ export class LuaTransformer {
...
+                        [tstl.createReturnStatement([itIdentifier])]
+                    )
+                )
+            ),
+            //return __it
+            tstl.createReturnStatement([itIdentifier]),
+        ];
+        return [block, functionScope];
+    }
     public transformFunctionDeclaration(functionDeclaration: ts.FunctionDeclaration): StatementVisitResult {
```
Comment (tomblind) [inline src/LuaTransformer.ts]: Can we get some line breaks between functions? 😄

File: src/TSHelper.ts [hunk 4]
```diff
@@ -178,6 +178,25 @@ export class TSHelper {
             return false;
         }
     }
+    public static isInGeneratorFunction(node: ts.Node, checker: ts.TypeChecker): boolean {
```
Comment (tomblind) [inline src/TSHelper.ts]: Is this used anywhere (maybe I'm not seeing it)?

File: src/TSHelper.ts [hunk 5]
```diff
@@ -178,6 +178,25 @@ export class TSHelper {
...
+            (n): n is ts.Node => ts.isFunctionDeclaration(n) || ts.isMethodDeclaration(n)
+        );
+        if (declaration && ts.isFunctionDeclaration(declaration) && declaration.asteriskToken) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    public static isGeneratorYieldCall(node: ts.Node, checker: ts.TypeChecker): boolean {
```
Comment (tomblind) [inline src/TSHelper.ts]: Also not seeing where this is used...

File: test/unit/functions.spec.ts [hunk 6]
```diff
@@ -388,4 +388,35 @@ export class FunctionTests {
         const result = util.transpileAndExecute(code);
         Expect(result).toBe("foobar");
     }
+    @Test("Generator functions")
+    public generatorFunction(): void {
```
Comment (tomblind) [inline test/unit/functions.spec.ts]: It seems like this should be broken into separate test functions.

File: src/LuaTransformer.ts (hunk 3 above)
Comment (andreiradu) [reply src/LuaTransformer.ts]: Done :)

File: src/LuaTransformer.ts [hunk 7]
```diff
@@ -1218,7 +1361,11 @@ export class LuaTransformer {
 
         return tstl.createExpressionStatement(this.transformExpression(expression));
     }
-
```
Comment (tomblind) [inline src/LuaTransformer.ts]: I know it seems minor, but spacing between functions helps a lot with readability

File: src/LuaTransformer.ts [hunk 8]
```diff
@@ -1218,7 +1361,11 @@ export class LuaTransformer {
...
+    public transformYield(expression: ts.YieldExpression): tstl.Expression {
+        return tstl.createCallExpression(
+            tstl.createTableIndexExpression(tstl.createIdentifier("coroutine"), tstl.createStringLiteral("yield")),
+                expression.expression?[this.transformExpression(expression.expression)]:[], expression);
+    }
     public transformReturn(statement: ts.ReturnStatement): tstl.Statement {
```
Comment (tomblind) [inline src/LuaTransformer.ts]: Same as above

File: src/LuaTransformer.ts (hunk 7 above)
Comment (andreiradu) [reply src/LuaTransformer.ts]: Done

**Pull request merged**